java-util/src/main/java/io/druid/java/util/common/CompressionUtils.java

```diff
@@ -102,8 +102,13 @@ public static long zip(File directory, OutputStream out) throws IOException
         totalSize += Files.asByteSource(file).copyTo(zipOut);
       }
       zipOut.closeEntry();
-      // Workarround for http://hg.openjdk.java.net/jdk8/jdk8/jdk/rev/759aa847dcaf
+      // Workaround for http://hg.openjdk.java.net/jdk8/jdk8/jdk/rev/759aa847dcaf
       zipOut.flush();
+
+      // fsync needs to happen before the FileOutputStream is closed by the try-with-resources
```

Comment:
This is going to get called in a few places where it isn't necessary (like the S3 pusher). It'd be better to only fsync when necessary. Add an option?

Reply:
we should probably move this logic to 
```
 public static long zip(File directory, File outputZipFile) throws IOException
```
method in the same class which passes FileOutputStream . Also add a `boolean fsync` flag to above method which would be true only for local pusher.

Reply:
@himanshug ah, I'm not sure if I was clear why I put it in this method instead of `public static long zip(File directory, File outputZipFile) throws IOException` and used the ugly instanceof - you can't sync on a FileOutputStream after it has been closed and the try-with-resources for the ZipOutputStream here closes the underlying FileOutputStream before the method returns, so trying to do it in `public static long zip(File directory, File outputZipFile) throws IOException` will result in a 'bad file descriptor' exception and will fail.

The two options I can think of are to do what I did and fsync before the streams are closed, or alternatively, I can remove the try-with-resources for ZipOutputStream and not call close on that stream. This might actually not be as bad an option as I first thought since ZipOutputStream.close() doesn't do much other than close the underlying stream and prevent new files from being written to the stream. In this case, it would be the caller's responsibility to close the stream they passed in which I think is perfectly reasonable (and which callers are already doing). What do you think?

Reply:
I see, I agree with the reasoning of not using try-with-resources and let callers be responsible for closing their streams.